### PR TITLE
crimson/osd/ops_executer: fix snap overlap range error

### PR DESCRIPTION
**Pull request by guojidan**
Fixes: https://tracker.ceph.com/issues/65113

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by guojidan** [comment]
now, `LibRadosSnapshotsSelfManagedPP, SnapOverlapPP` test case can be passed through

This bug is due to the lack of statistics on the `modified_ranges` of each write op

File: src/crimson/osd/ops_executer.cc [hunk 1]
```diff
@@ -970,20 +971,28 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
...
+      interval_set<uint64_t> &newest_overlap =
+      cloning_ctx->new_snapset.clone_overlap.rbegin()->second;
+      modified_ranges.intersection_of(newest_overlap);
+      newest_overlap.subtract(modified_ranges);
+    } else if (!obc->ssc->snapset.clones.empty()) {
+      interval_set<uint64_t> &newest_overlap =
+      obc->ssc->snapset.clone_overlap.rbegin()->second;
+      modified_ranges.intersection_of(newest_overlap);
+      newest_overlap.subtract(modified_ranges);
+    }
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.cc:992]: Boilerplate can be avoided here:
```
   interval_set<uint64_t> &newest_overlap;
   if (cloning_ctx) {
      newest_overlap =
      cloning_ctx->new_snapset.clone_overlap.rbegin()->second;
    } else if (!obc->ssc->snapset.clones.empty()) {
      newest_overlap =
      obc->ssc->snapset.clone_overlap.rbegin()->second;
    }
    modified_ranges.intersection_of(newest_overlap);
    newest_overlap.subtract(modified_ranges);
```

File: src/crimson/osd/ops_executer.cc [hunk 2]
```diff
@@ -970,20 +971,28 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
...
-  // was just created (See prepare_clone()).
-  interval_set<uint64_t> &newest_overlap =
-    cloning_ctx->new_snapset.clone_overlap.rbegin()->second;
-  osd_op_params->modified_ranges.intersection_of(newest_overlap);
-  delta_stats.num_bytes += osd_op_params->modified_ranges.size();
-  newest_overlap.subtract(osd_op_params->modified_ranges);
   return cloning_ctx;
 }
 
+void OpsExecuter::update_clone_overlap(OSDOp& osd_op) {
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.cc:977]: The removed bit had `delta_stats.num_byte` usage, was it omitted intentionally?

**Review by Matan-B — COMMENTED** [review]
> now, LibRadosSnapshotsSelfManagedPP, SnapOverlapPP test case can be passed through

@guojidan Cool, great catch!
Left some comments.

**Comment by guojidan** [comment]
hi @Matan-B , Please take a look at my current code, I think this is more reasonable

File: src/crimson/osd/pg_backend.cc [hunk 3]
```diff
@@ -593,14 +593,12 @@ void PGBackend::update_size_and_usage(object_stat_sum_t& delta_stats,
...
-    if (oi.size) {
-      ch.insert(0, oi.size);
-    } else if (length) {
-      ch.insert(offset, length);
-    }
-    modified.union_of(ch);
+  if (write_full && oi.size) {
+    ch.insert(0, oi.size);
+  } else {
+    ch.insert(offset, length);
```
Comment (Matan-B) [inline src/crimson/osd/pg_backend.cc:599]: This is wrong on write_full=true and oi.size=0, I think it should be:
```
  if (write_full) {
    if (oi.size) {
      ch.insert(0, oi.size);
    }
  } else if (length) {
    ch.insert(offset, length);
  }
  modified.union_of(ch);
```

File: src/crimson/osd/ops_executer.cc [hunk 4]
```diff
@@ -970,20 +970,28 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
...
-  interval_set<uint64_t> &newest_overlap =
-    cloning_ctx->new_snapset.clone_overlap.rbegin()->second;
-  osd_op_params->modified_ranges.intersection_of(newest_overlap);
-  delta_stats.num_bytes += osd_op_params->modified_ranges.size();
-  newest_overlap.subtract(osd_op_params->modified_ranges);
   return cloning_ctx;
 }
 
+void OpsExecuter::update_clone_overlap() {
+  if (op_info.may_write() && obc->obs.exists && !snapc.snaps.empty()) {
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.cc:977]: This condition is equal to `op_info.may_write() && should_clone(*obc, snapc)` which is already checked in OpsExecuter's ctor. If the condition is true, `cloning_ctx` should have a value (non-null).
`OpsExecuter::flush_clone_metadata` checks `if (cloning_ctx)` - IIUC we should call `update_clone_overlap` there.

File: src/crimson/osd/ops_executer.cc [hunk 5]
```diff
@@ -970,20 +970,28 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
...
+    if (cloning_ctx) {
+      newest_overlap =
+      &cloning_ctx->new_snapset.clone_overlap.rbegin()->second;
+    } else if (!obc->ssc->snapset.clones.empty()) {
+      newest_overlap =
+      &obc->ssc->snapset.clone_overlap.rbegin()->second;
+    } else {
+      return;
+    }
+    if (osd_op_params) {
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.cc:988]: Moreover, if we have `cloning_ctx` value, then `want_mutate` should also be true. Meaning we must also have have `osd_op_params`. We should be able to assert `osd_op_params` here.

File: src/crimson/osd/ops_executer.cc [hunk 6]
```diff
@@ -974,21 +973,21 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
   return cloning_ctx;
 }
 
-void OpsExecuter::update_clone_overlap(OSDOp& osd_op) {
+void OpsExecuter::update_clone_overlap() {
   if (op_info.may_write() && obc->obs.exists && !snapc.snaps.empty()) {
-    interval_set<uint64_t> modified_ranges;
-    modified_ranges.insert(osd_op.op.extent.offset, osd_op.op.extent.length);
-    logger().debug("{} modified_ranges: {}", __func__, modified_ranges);
+    interval_set<uint64_t> *newest_overlap;
```
[the remark this answers is not in the document]
Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:978]: reference will throw error，`Declaration of reference variable 'newest_overlap' requires an initializer`

[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc:978]: Please see the last comments, if they are correct, we may have a single newest_overlap initialization.

File: src/crimson/osd/ops_executer.cc (hunk 5 above)
Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:988]: `cloning_ctx` is non-null requires `snapc.snaps[0] > initial_obc.ssc->snapset.seq` is true, but when write op occur and `snapc.snaps[0] > initial_obc.ssc->snapset.seq` is false like [source](https://github.com/ceph/ceph/blob/main/src/test/librados/snapshots_cxx.cc#L283), `cloning_ctx` will be null, but we also need update `clone_overlap`

File: src/crimson/osd/ops_executer.cc (hunk 4 above)
Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:977]: this condition not equal to `op_info.may_write() && should_clone(*obc, snapc)`, this condition not contains `snapc.snaps[0] > initial_obc.ssc->snapset.seq`

File: src/crimson/osd/ops_executer.cc (hunk 5 above)
Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:988]: yes,  We should be able to assert osd_op_params here, because write op always initialize `osd_op_params`

**Review by Matan-B — COMMENTED** [review]
LGTM, please squash the last two commits

**Comment by guojidan** [comment]
> LGTM, please squash the last two commits

sorry, I squash all the commits into one

**Review by athanatos — APPROVED** [review]

File: src/crimson/osd/ops_executer.cc [hunk 7]
```diff
@@ -970,20 +970,29 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
...
+    && !obc->ssc->snapset.clones.empty()) {
+    newest_overlap =
+      &obc->ssc->snapset.clone_overlap.rbegin()->second;
+  } else {
+    return;
+  }
+
+  assert(osd_op_params);
+  osd_op_params->modified_ranges.intersection_of(*newest_overlap);
+  newest_overlap->subtract(osd_op_params->modified_ranges);
```
Comment (xxhdx1985126) [inline src/crimson/osd/ops_executer.cc:993]: Seems that the line `delta_stats.num_bytes += osd_op_params->modified_ranges.size();` should be kept, am I missing something?

Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:993]: maybe already processed this in `PGBackend::update_size_and_usage` , [here](https://github.com/ceph/ceph/blob/main/src/crimson/osd/pg_backend.cc#L607)

Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc:993]: The clone object size won't be added `PGBackend::update_size_and_usage`.
Can you please add the missing line?

Comment (guojidan) [reply src/crimson/osd/ops_executer.cc:993]: wow yes, Both the head object and clone object need to calculate the modified size

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-snap-overlap

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by guojidan** [comment]
thanks all 😄

**Pull request merged**